src/critnib/critnib.c

```diff
@@ -290,8 +315,20 @@ static struct critnib_leaf *alloc_leaf(struct critnib *__restrict c) {
     }
 
     struct critnib_leaf *k = c->deleted_leaf;
-
     c->deleted_leaf = k->value;
+
+    uint64_t ref_count;
+    void *to_be_freed;
+    void *key;
+    utils_atomic_load_acquire_u64(&k->ref_count, &ref_count);
+    utils_atomic_load_acquire_ptr(&k->to_be_freed, &to_be_freed);
+    utils_atomic_load_acquire_ptr((void **)&k->key, &key);
+    if (ref_count && to_be_freed) {
+        LOG_DEBUG("WARNING: reusing probably non-freed leaf: key = %p, value = "
```

Comment:
should we also put an assert here that would trigger in debug mode?

Reply:
I could but I am not sure if we should. 1) It is extremely unlikely and 2) it might cause only a memory leak of one value that IMHO is a good trade-off versus further complicating and decreasing performance of critnib. What do you think?

Reply:
Hmm ... yes, right, it may lead to UB, so even a crash cannot be excluded.
I added an assert here.
Done.